content/posts/2020/04/libplanet-0.9.eng.md

```diff
@@ -0,0 +1,117 @@
+
+[Libplanet.RocksDBStore]: https://www.nuget.org/packages/Libplanet.RocksDBStore/
+[IStore]: https://docs.libplanet.io/0.9.0/api/Libplanet.Store.IStore.html
+[RocksDB]: https://rocksdb.org/
+[DefaultStore]: https://docs.libplanet.io/0.9.0/api/Libplanet.Store.DefaultStore.html
+[1]: {{< ref "rocksdb/index.eng.md" >}}
+
+
+Receiving Blocks from Multiple Peers
+------------------------------------
+
+Up to this version, [`PreloadAsync()`][Swarm.PreloadAsync] and [`StartAsync()` Method][Swarm.StartAsync] of [`Swarm<T>` Class][Swarm] requested and received all blocks from a single peer to catch up with blocks piled up on the network. But unfortunately, if there were many piles of blocks, it would take a long time to get them from just one peer.
+And if you were unlucky, you might request blocks from a peer with a very slow connection which would take extra longer time. It also put a lot of pressure on the peer that’s sending the blocks and it was especially hard to ignore the burden for seed nodes that were set by default for deployed applications.
+
+So starting with this version, block downloads are improved to be evenly distributed across multiple peers and even when there’s a slow peer among the senders, the downloading speed has been significantly reduced. 
+
+
+[Swarm]: https://docs.libplanet.io/0.9.0/api/Libplanet.Net.Swarm-1.html
+[Swarm.PreloadAsync]: https://docs.libplanet.io/0.9.0/api/Libplanet.Net.Swarm-1.html#Libplanet_Net_Swarm_1_PreloadAsync_System_Nullable_TimeSpan__IProgress_Libplanet_Net_PreloadState__IImmutableSet_Libplanet_Address__EventHandler_Libplanet_Net_PreloadBlockDownloadFailEventArgs__CancellationToken_
+[Swarm.StartAsync]: https://docs.libplanet.io/0.9.0/api/Libplanet.Net.Swarm-1.html#Libplanet_Net_Swarm_1_StartAsync_TimeSpan_TimeSpan_CancellationToken_
+
+
+Signed App Protocol Version
+---------------------------
+
+Last year, [Libplanet 0.3 added `appProtocolVersion` parameter in the `Swarm<T`() constructor.][2] This allowed nodes with compatible protocols to communicate with each other and adequately handled nodes without compatible protocols according to the application.
+
+Our team also took advantage of this feature and used it for software updates when encountering a higher version of the node. But with this use, we realized that a modulated software could be used to maliciously display high-version numbers (which have never been reported) and attack other nodes to attempt false software updates.
+
+To avoid this, the application protocol version that used to be described as [`System.Int32`][System.Int32] is now changed to [`AppProtocolVersion` Data Type][AppProtocolVersion], which includes multiple metadata such as signatures and signers. The app protocol version must be signed, and each node will *individually* determine which signer's version of the app protocol it will trust using the `trustedAppProtocolVersionSigners` parameter in the `Swarm<T>()` constructor.
```

Comment:
```suggestion
To avoid this, the application protocol version that used to be described as [`System.Int32`][System.Int32] is now changed to [`AppProtocolVersion` type][AppProtocolVersion], which includes multiple metadata such as signatures and signers. The app protocol version must be signed, and each node will *individually* determine which signer's version of the app protocol it will trust using the `trustedAppProtocolVersionSigners` parameter in the `Swarm<T>()` constructor.
```